Rework java logging and trace correlation (#9717)
DataDog/documentation: content/en/logs/log_collection/java.md

{{% /tab %}}
{{% tab "Log4j2" %}}

log4j2 includes a JSON layout. Configure a file appender using the JSON layout in `log4j2.xml`:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<Configuration>
  <Appenders>
    <File name="FILE" fileName="logs/app.log" >
      <JSONLayout compact="true" eventEol="true" properties="true" stacktraceAsString="true" />
    </File>
  </Appenders>

  <Loggers>
    <Root level="INFO">
      <AppenderRef ref="FILE"/>
    </Root>
  </Loggers>
</Configuration>
```

Then, add the JSON layout dependencies to your `pom.xml`:

```xml
<dependency>
    <groupId>org.apache.logging.log4j</groupId>
    <artifactId>log4j-core</artifactId>
    <version>2.7</version>
</dependency>
<dependency>
    <groupId>com.fasterxml.jackson.core</groupId>
    <artifactId>jackson-core</artifactId>
    <version>2.8.3</version>
</dependency>
<dependency>
    <groupId>com.fasterxml.jackson.core</groupId>
    <artifactId>jackson-databind</artifactId>
    <version>2.8.3</version>
</dependency>
<dependency>
    <groupId>com.fasterxml.jackson.core</groupId>
    <artifactId>jackson-annotations</artifactId>
    <version>2.8.3</version>
</dependency>
```

{{% /tab %}}
{{% tab "Logback" %}}

> We recommend using the [logstash-logback-encoder][1] for JSON formatted logs in Logback. Configure a file appender using the JSON layout in `logback.xml`:

Reviewer: ```suggestion
You can use the [logstash-logback-encoder][1] for JSON formatted logs in Logback. To configure a file appender using the JSON layout in `logback.xml`:
```

```xml
<configuration>
  <appender name="FILE" class="ch.qos.logback.core.FileAppender">
    <file>logs/app.log</file>
    <encoder class="net.logstash.logback.encoder.LogstashEncoder" />
  </appender>

  <root level="INFO">
    <appender-ref ref="FILE"/>
  </root>
</configuration>
```

Then, add the logstash encoder dependency to your `pom.xml` file:

```xml
<dependency>
  <groupId>ch.qos.logback</groupId>
  <artifactId>logback-classic</artifactId>
  <version>1.1.3</version>
</dependency>
<dependency>
  <groupId>net.logstash.logback</groupId>
  <artifactId>logstash-logback-encoder</artifactId>
  <version>4.5.1</version>
</dependency>
```

[1]: https://github.com/logstash/logstash-logback-encoder
{{% /tab %}}
{{< /tabs >}}

**Inject trace IDs in your logs**

> If APM is enabled for this application, you can correlate logs and traces by [following these instructions][6] to enable trace id injection.

Reviewer: ```suggestion
If APM is enabled for this application, you can correlate logs and traces by referring to [Connecting Java Logs and Traces][6] to enable trace id injection.
```

There are two steps to configure the Java application to stream logs directly to Datadog:

1. Add the Logback logging library to your code (or build a bridge from your current logger to it)
2. Configure Logback to send logs to Datadog

### Bridge from Java logging libraries to Logback

> Most common logging libraries can be bridged to Logback

Reviewer: What does this mean, can you include more context? Also looks like this needs a period to be a complete statement.

Author: This is connected to the "There are two steps to configure the Java application to..." section right before.

How about:

```suggestion
If you are not using Logback as your logging library, the first step to stream logs directly to Datadog is to bridge your logging library to Logback. The exact steps depends on your logging library:
```

{{< tabs >}}
{{% tab "Log4j" %}}

> We advise you to use an slf4j module called [log4j-over-slf4j][1] along with Logback.

Reviewer: ```suggestion
Logging to a remote server in JSON may be difficult. You can use the slf4j module [log4j-over-slf4j][1] with Logback to simplify this process.
```

1. Replace your `log4j.jar` dependency with a `log4j-over-slf4j.jar` dependency
2. Add the logback dependencies
3. Configure logback

Your `pom.xml` file should have the following dependencies:

```xml
<dependency>
  <groupId>org.slf4j</groupId>
  <artifactId>log4j-over-slf4j</artifactId>
  <version>1.7.13</version>
</dependency>
<dependency>
  <groupId>ch.qos.logback</groupId>
  <artifactId>logback-classic</artifactId>
  <version>1.1.3</version>
</dependency>
<dependency>
  <groupId>net.logstash.logback</groupId>
  <artifactId>logstash-logback-encoder</artifactId>
  <version>4.5.1</version>
</dependency>
```

**note:** As a result of this migration, Log4j configuration files will no longer be picked up. Migrate your `log4j.properties` file to `logback.xml` with the [Log4j translator][1].

[1]: http://logback.qos.ch/translator
[2]: http://www.slf4j.org/legacy.html#log4j-over-slf4j
{{% /tab %}}

{{% tab "Log4j2" %}}

Log4j2 allows logging to a remote host, but it does not offer the ability to prefix the logs by an API key. Because of this, it is recommended that you use a SLF4J  module called `log4j-over-slf4j` and Logback.

> `log4j-to-slf4j.jar` seamlessly replaces log4j2 in your application so you do not have to make any code changes.  To use it

Reviewer: ```suggestion
`log4j-to-slf4j.jar` seamlessly replaces log4j2 in your application so you do not have to make any code changes. To use it:
```

            </layout>
          </prefix>
    </encoder>
  </appender>

  <root level="DEBUG">
    <appender-ref ref="FILE"/>
    <appender-ref ref="JSON_TCP" />
  </root>
</configuration>
```

{{< /site-region >}}
{{< site-region region="eu" >}}

```xml
<configuration>
  <appender name="FILE" class="ch.qos.logback.core.FileAppender">
    <file>logs/app.log</file>
    <encoder class="net.logstash.logback.encoder.LogstashEncoder" />
  </appender>
  <appender name="JSON_TCP" class="net.logstash.logback.appender.LogstashTcpSocketAppender">
    <remoteHost>tcp-intake.logs.datadoghq.eu</remoteHost>
    <port>1883</port>
    <keepAliveDuration>20 seconds</keepAliveDuration>
    <encoder class="net.logstash.logback.encoder.LogstashEncoder">
        <prefix class="ch.qos.logback.core.encoder.LayoutWrappingEncoder">
            <layout class="ch.qos.logback.classic.PatternLayout">

>                 <pattern><APIKEY> %mdc{keyThatDoesNotExist}</pattern>

Author: ```suggestion
                <pattern><API_KEY> %mdc{keyThatDoesNotExist}</pattern>
```
